### PR TITLE
[PERF] list object with partial response support

### LINKED ISSUE #88: [Perf] Partial Response object listing support
Support the partial response support https://cloud.google.com/storage/docs/json_api#partial-response to speed up.

For a bucket which contains more than 60K 
- Without optim: 1m40s
- WIth partial response: 40s

### PR DESCRIPTION
**Pull request by cboudereau**
Hello.

Fixes #88 

I would like to contribute and discuss to support partial response object from gcs listing in order to speed up the listing on large bucket/prefix. 

Here is the documentation explaining the performance optimization: https://cloud.google.com/storage/docs/json_api#partial-response

FYI I tried with a bucket which contains more than 60K objects. It took 1m40s with full object response rather than 40s with only the selfLink field in Object.

To test it, in the example, you can comment in/out the line which contains the new field option on the request parameter

Please, consider this PR as an experiment to open the discussion and see how we can integrate partial response support.

Sorry for the crappy version but I would like to bench first this implementation to replace our custom implementation.

Thanks in advance

### CODE REVIEW

**Comment by ThouCheese** [comment]
So how could we design this is a way that is ergonomic and safe to use? From what I understand the `fields` parameter controls which fields are populated in the JSON response sent by the GCS API. Then a 'general` response would need to have each of its fields be an option when the fields parameter is supported. This would essentially mean opting out of Rust's nice null-safety guarantees, i.e. the programmer has to make sure that when they access a field on `cloud_storage` struct, they also include that field in the `fields` parameter. Do you have any thoughts on what a convenient and safe API for this feature could look like?

**Comment by cboudereau** [comment]
Hello.

Thank you for your quick feedback. I tried in the first version to make
everything optional in the Object and ListObject but we loose some
mandatory information like bucket for instance.

Let me know if you have an idea in order integrate the feature. I will try
to find out a solution for the coming week.

By the way do not hesitate to add commit or though in the PR 🙂

**Comment by cboudereau** [comment]
Hello @ThouCheese. The PR is ready, I just left name, bucket and self_link mandatory and the performance is still improved as if there was only self_link. That way, we have the performance optimization without breaking all the API.

File: Cargo.toml [hunk 1]
```diff
@@ -38,4 +38,4 @@ futures =          { version = "0.3",  default_features = false, features = ["al
 bytes =            { version = "1.0",  default-features = false }
 
 [package.metadata.docs.rs]
-features = ["global-client", "sync"]
+features = ["global-client", "sync"]
```
Comment (ThouCheese) [inline Cargo.toml:41]: Newline is missing here.

File: src/resources/object.rs [hunk 2]
```diff
@@ -147,6 +147,13 @@ pub struct ObjectPrecondition {
 #[derive(Debug, PartialEq, serde::Serialize, Default, Clone)]
 #[serde(rename_all = "camelCase")]
 pub struct ListRequest {
+    /// When specified, returns only specified field.
+    /// See [the Google Cloud Storage Partial Response](https://cloud.google.com/storage/docs/json_api)
+    /// ```
+    /// let fields = "items(selfLink,name,bucket),nextPageToken";
```
Comment (ThouCheese) [inline src/resources/object.rs:153]: What does this example mean?

**Review by ThouCheese — CHANGES_REQUESTED** [review]

**Comment by ThouCheese** [comment]
Hi @cboudereau, I have left some comments behind on the changes, but I also don't see how you addressed the question I raised earlier about the safety of the API. It looks like you converted most fields to `Option<T>`, but this still means that type information gets erased. If we cannot come up with something better I suggest we create `PartialObject`, `PartialListRequest` and `Object::list_partial`, and then use the less type-safe version there. That way is it more opt-in.

File: src/resources/object.rs (hunk 2 above)
Comment (cboudereau) [reply src/resources/object.rs]: done

**Comment by ThouCheese** [comment]
Hi! Getting close, but a couple of remarks:

1. `list_partial` is still missing from the sync client,
2. The docs for `list_partial` and `PartialObject` should tell the user why they are different from `list` and `Object`, and also `ListRequest` still says that the struct is used for `Object::list`,
3. Do you think that `list_request.fields` could be a Vec of some enum, say `ObjectField`, so that the user cannot make typos when listing the object? That could be a nice extra safe api.
4. There should be tests for these new functions.

File: src/client/object.rs [hunk 3]
```diff
@@ -145,6 +150,43 @@ impl<'a> ObjectClient<'a> {
         bucket: &'a str,
         list_request: ListRequest,
     ) -> crate::Result<impl Stream<Item = crate::Result<ObjectList>> + 'a> {
+        let stream = self.list_objects(bucket, list_request).await?;
+        Ok(stream.map(|x| x?.try_into().map_err(crate::error::Error::Other)))
+    }
+
+    /// Obtain a list of objects within this Bucket.
```
Comment (ThouCheese) [inline src/client/object.rs:157]: This should denote that this is only the partial response.

File: src/resources/object.rs [hunk 4]
```diff
@@ -95,6 +95,95 @@ pub struct Object {
     pub kms_key_name: Option<String>,
 }
 
+/// A resource representing a file in Google Cloud Storage with PartialResponse support.
```
Comment (ThouCheese) [inline src/resources/object.rs:98]: This comment should explain why it is different from the `Object`.

**Review by ThouCheese — CHANGES_REQUESTED** [review]

**Comment by cboudereau** [comment]
> Hi! Getting close, but a couple of remarks:
> 
> 1. `list_partial` is still missing from the sync client,
done
> 2. The docs for `list_partial` and `PartialObject` should tell the user why they are different from `list` and `Object`, and also `ListRequest` still says that the struct is used for `Object::list`,
done
> 3. Do you think that `list_request.fields` could be a Vec of some enum, say `ObjectField`, so that the user cannot make typos when listing the object? That could be a nice extra safe api.
I don't think so since it is more like a query and not used anywhere else
> 4. There should be tests for these new functions.
done

**Pull request closed without merging**